### PR TITLE
fix(enhance): preserve CLAHE clipping equivariance

### LINKED ISSUE #4632: equalize_clahe slow path is not transposition-equivariant below float64 when clipping
On the differentiable path with a clip limit, `equalize_clahe` is not equivariant under transposing the
image and the grid — even for a **square** grid, where no axis asymmetry is involved. The output moves by
up to `2.9e-2` in `bfloat16`.

## Reproduction

```python
import torch
from kornia import enhance

for dt in (torch.float64, torch.float32, torch.bfloat16, torch.float16):
    bad, worst = 0, 0.0
    for s in range(50):
        torch.manual_seed(s)
        img = torch.rand(1, 1, 12, 12, dtype=dt)
        a = enhance.equalize_clahe(img, 40.0, (2, 2), True)
        b = enhance.equalize_clahe(img.transpose(-2, -1), 40.0, (2, 2), True).transpose(-2, -1)
        d = (a.float() - b.float()).abs().max().item()
        worst = max(worst, d)
        bad += d > 1e-3
    print(dt, bad, worst)
```

Measured on `main` (`3f748920e`), torch 2.14.0, CPU, macOS arm64:

| dtype | seeds past `1e-3` | worst abs diff |
| --- | --- | --- |
| float64 | 0/50 | `0.000e+00` |
| float32 | 0/50 | `1.192e-07` |
| **bfloat16** | **48/50** | **`2.930e-02`** |
| float16 | 0/50 | `9.766e-04` |

float64 is exact and float32 is at the rounding floor, so this is precision-dependent rather than an
indexing bug. bfloat16 is affected on nearly every seed; float16 sits just under the threshold but its
worst case is already `9.8e-4`, i.e. the same phenomenon a factor of 30 smaller.

## Likely cause

The clip-and-redistribute step sums a soft histogram and then treats the excess as a count. Below
float64 a rounding residue in that sum can cross an integer boundary, which changes how much mass is
redistributed — and the residue depends on summation order, which transposition changes. The fast path
does not show it because it histograms with `bincount` on integers.

Not a regression: identical on `main` and on #4628's merge tree. Found while reviewing #4628, which
adds a transposition test — that test deliberately runs the slow path at `clip_limit=0` to avoid this,
and the reason should be this issue rather than a comment.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

### PR DESCRIPTION
**Pull request by NiteeshGowda72**
## What does this pull request do?

This pull request fixes a numerical issue in the slow differentiable path of `equalize_clahe` that could cause the output to depend on the spatial ordering of the input when histogram clipping was enabled.

### Problem

In `_compute_luts()`, the clipped histogram mass was previously calculated indirectly:

histos.clamp_(max=max_val)
clipped = pixels - histos.sum(1)

For the slow differentiable path, the histogram is represented using floating-point values. In low-precision dtypes, the reduction performed by histos.sum(1) can introduce small rounding errors.

Because floating-point reduction can depend on the order in which values are accumulated, transposing the input can change the rounding error. This affects the subsequent redistribution of the clipped histogram mass.

In some cases, this numerical error could change the residual redistribution and produce different CLAHE outputs for an image and its transposed version.

Solution

Instead of deriving the clipped mass from the total pixel count after clamping, this change computes the actual amount exceeding the clipping limit before modifying the histogram:

clipped = torch.relu(histos - max_val).sum(1)
histos.clamp_(max=max_val)

This directly measures the histogram mass that needs to be redistributed and avoids deriving it from the floating-point difference:

pixels - histos.sum(1)

The existing redistribution logic then operates on the explicitly computed clipped mass.

Tests

Added a regression test covering transpose equivariance for the slow differentiable CLAHE path with:

clip_limit=40.0
grid_size=(2, 2)
slow_and_differentiable=True
float32 input
deterministic input using torch.manual_seed(28)

The test compares the normal output with the output obtained by transposing the input and transposing the result back.

Changelog

Added a changelog entry describing the numerical fix in the slow differentiable CLAHE path.

Related issue

Closes #4632

What did you check?
Focused tests
pixi run -e cuda test-module tests/enhance/test_equalization.py

26 passed, 1 skipped
Lint
pixi run -e cuda lint

ruff check: Passed
ruff format: Passed
Type checking
pixi run -e cuda typecheck

No type errors
Full test suite
pixi run -e cuda test

11,367 passed
3,742 skipped
44 xfailed
11 xpassed
275 warnings

No test failures were observed.

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
`REQUEST_CHANGES` on `8cbfd001c6d0b092c3c31c73e2d72ef53ec9ef89`.

**[P1] Restrict the new clipped-mass formula to the differentiable path.** The assignment at `kornia/enhance/equalization.py:176` also changes the fast histogram path. CPU `histc` skips marginally out-of-range samples, and the old `pixels - histos.sum(1)` deliberately redistributed those missing samples after clamping; `relu(histos - max_val).sum(1)` drops them from the LUT mass instead. This input is in the tolerance window that `RandomClahe` documents and #4564 preserves. On both the reviewed head and the semantic merge with #4611, the following accepted fast-path case changes materially:

```python
x = torch.linspace(0, 1, 64).reshape(1, 1, 8, 8)
x.view(-1)[0] = 1.0000001
y = equalize_clahe(x, 40.0, (1, 1))
```

The base output sum/max are `32.878433 / 1.0`; this head returns `31.874510 / 0.984314`. With eight such samples the sum changes `39.356865 -> 31.325493`; with `-1e-7`, the first mapped value changes `0.011765 -> 0`. This also invalidates #4611's exact CPU-preservation contract, while this PR's fragment describes only slow-path numerical errors. Preserve `pixels - histos.sum(1)` for the discrete fast path and use the excess-only formula only for `diff=True`, with an end-to-end out-of-range pin.

**[P2] The new regression passes unchanged on the buggy base.** `tests/enhance/test_equalization.py:341` hardcodes CPU float32, although #4632 is a reduced-precision failure (bfloat16 reached `2.93e-2`; float32 was already at its rounding floor). Re-running the added seed-28 assertion against base `607869e2d` gives a maximum difference of `5.96e-08`, and `self.assert_close(..., low_tolerance=True)` passes. The test therefore cannot catch reverting the production line this PR changes.

Please pin the affected regime. For example, seed 11 with CPU bfloat16 measures `2.9296875e-2` on the base and `3.90625e-3` on this head, so the normal bfloat16 `BaseTester` tolerance distinguishes the fix. Better still, remove the `clip_limit = 0.0 if slow_and_differentiable else 40.0` workaround in the existing fixture-driven transpose test at `tests/enhance/test_equalization.py:333`, so clipping is exercised across the supported device/dtype matrix and non-square grids too.

For the differentiable path, the formula is mathematically sound: it redistributes only mass actually above the clipping limit instead of treating soft-histogram normalization roundoff as clipped counts. Focused validation on the head passed (`130 passed, 4 skipped` for the equalization file across CPU float16, bfloat16, float32, and float64). A 50-seed comparison confirms the bfloat16 worst case drops from `2.9296875e-2` to one bfloat16 step, `3.90625e-3`; float16/float32/float64 remain at their rounding floors.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5.6-sol).

**Comment by NiteeshGowda72** [comment]
Thanks for the review. I’ve addressed the feedback in the latest commit.

Changes made:

Updated the CLAHE clipping logic so the slow differentiable path (diff=True) uses the excess histogram mass when redistributing clipped bins, which preserves transpose equivariance below float64.
Preserved the existing fast histogram path (diff=False) behavior using pixels - histos.sum(1) after clipping, including the handling of marginally out-of-range input values.
Updated the existing transpose-equivariance test to exercise clipping on both fast and slow differentiable paths across non-square grid configurations.
Added a regression test covering the out-of-range fast-path behavior.
Kept the dedicated regression test for #4632.
Added the requested changelog entry.

Validation:

tests/enhance/: 404 passed, 15 skipped
ruff-check: passed
ruff-format: passed
typecheck: no errors
git diff --check: clean

The changes are pushed in commit 58752344.

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
Review of `58752344c0cc56c4a9570e3ef4e3746314991bb5` — **REQUEST_CHANGES**.

**[P2] Remove the duplicate changelog fragment** (`changelog.d/4579.fixed.md:1`). This head now adds both `4579.fixed.md` and `4637.fixed.md` for the same CLAHE fix, so the release notes will announce it twice and one entry is attributed to an unrelated PR number. Keep the more complete text under `4637.fixed.md` and delete `4579.fixed.md`.

The two substantive blockers from the previous review are fixed: the excess-only clipped-mass calculation is restricted to the differentiable path, the fast path retains `pixels - histos.sum(1)`, and the fixture-driven transpose test now exercises clipping across devices, dtypes, and non-square grids. The focused CPU float16/bfloat16/float32/float64 equalization run is green: 131 passed, 4 skipped. All 41 reported checks are green.

This head also conflicts with current `main` in `tests/enhance/test_equalization.py`; that conflict must be resolved and the resulting merge tested before it can merge.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-5.6-sol).

**Review by ducha-aiki — APPROVED** [review]
APPROVE on `7aefb5db52de0dffd8a6612a5a7a4a6588841653`.

Addressed the remaining review feedback directly and pushed a merge commit without rewriting branch history. Merged main `8ca60bc34`, resolved the equalization-test conflict by keeping both the PR's fast-path regression and main's single-tile, histogram and compile regressions, and removed the duplicate `4579.fixed.md` fragment. The fuller release note is retained under `4637.fixed.md`.

Validation on the resulting tree: `pixi run -e default test-module tests/enhance/test_equalization.py --device=cpu --dtype=float16,bfloat16,float32,float64` passed 161 tests with 6 skipped; the two dynamo cases passed with `KORNIA_TEST_OPTIMIZER=inductor`; full `pixi run -e default pre-commit-all` passed after completing the merge. Imports were verified to resolve to the isolated checkout. CUDA/MPS and the full test suite were not rerun for this fix.

Both blocking findings from my previous review are resolved. GitHub confirms the pushed head is mergeable; fresh CI remains the merge gate.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-6).
